### ISSUE TITLE
Accessing schemas defined in YAML within Fastify app

### ISSUE THREAD

**Issue by jesselathamdev** [seriousme/fastify-openapi-glue#305]
Hi there, I'm working on understanding the best use of fastify-openapi-glue with respect to accessing the schemas defined within the OpenAPI document.

Mainly my digging has lead me to believe that as the schemas are defined in the YML file, and that file is loaded at app start, I would think that the schemas should all be available through fastify.getSchemas() which would then allow me to use them in part to further validate the incoming data with custom validation (ex: validation of multiple fields, searching the database for unique values, and so on).  

However, when I examine fastify.getSchemas() this appears to be empty.  Is there another way to access the schemas as they are loaded from glue so that I can use Ajv ahead of time to set up the validations where I would simply pass the request to?

In addition to the above, I would love to have the data available in a prepared POJO (like a global fastify schema) that that has been filtered which I could then pass around into other functions or methods instead of using request.body which may contain data that hasn't been treated. The alternative here is that I have to write a POJO and fill it myself from the request.body to then be able to use as I see fit.

Is there a best way from your experience to load schemas into the global collection via fastify.addSchemas?

Thoughts and thanks for your time in advance on this question.

**Comment by seriousme** [comment]
Hi, thanks for asking.

Not sure I understand you correctly though. 

The idea behind fastify-openapi-glue is that you give it an [OpenAPI](https://spec.openapis.org/oas/v3.1.0) schema (not a fastify schema) and it will generate a fastify _configuration_ (including routes and schema) for you on the fly so you do not have to bother with schema validation yourself as fastify will take care of this for you.  Eg. see the [petstore example](https://github.com/seriousme/fastify-openapi-glue/blob/v2.x/examples/petstore/index.js) which you can run by executing `fastify start index.js`.

If you want full access to the schema and/or the request.body then it might be more convenient to use fastify without fastify-openapi-glue.

Hope this answers your question.

Kind regards,
Hans

**Comment by jesselathamdev** [comment]
Hi Hans,

Thanks for the timely response!

For my case I'm thinking of an escape hatch where I'd like the capabilities that fastify-openapi-glue ships with, but to also populate the global schemas so that I have an escape hatch for advanced validation cases and to do more with the data than basic CRUD.

An example here would be to pass the schema validated data and filled object into a complex "form/processor handler" which I would then do a number of things with such as: additional validation beyond simple schema, various database operations touching multiple tables which belong in a transaction, trigger notifications and other background jobs from, and so on.  The encapsulation here would allow us to keep those types of concerns together and make it much more testable than having all this code inside a fastify request definition (we hand off the processing).

I've inspected the fastify request object and can see the context.schema details there, but would like know if we can have that registered in advance to work with, ideally immediately available in global schemas, or as a whitelisted set of schemas that are allowed to be made global.

In summary, I want to keep with the spirit of "glue" where the schema is defined, but I would like to be able to access the schema outside of the request/response to help simplify the architecture of a growing applications surface area.

Is this something you could imagine adding (and doable) as an option to glue to expose the schemas we've already defined in the OpenAPI spec so that we don't have to write them again?

Cheers and thanks,
Jesse

**Comment by seriousme** [comment]
Jesse,

I think I better understand now what you mean ;-)

The whole idea of glue is to abstract the whole OpenApi and most of the Fastify stuff so devs can focus on service.js (and security.js ;-)).  Exposing the details would break that abstraction.

Fastify itselfs [encapsulates](https://www.fastify.io/docs/latest/Reference/Encapsulation/) the config of the plugin, so thats why you can't see the config of the plugin from the parent instance.

There is howver a plugin called [fastify-routes](https://github.com/fastify/fastify-routes) that should give you full access to any config added by plugins. I have not tried it myself but it looks like that might be able to help you out.

Kind regards,
Hans
